docs/DEX.md

```diff
@@ -0,0 +1,115 @@
+# DEX
+
+## Name
+When I first tried to write this doc, I kept on saying stuff like "the data system" or "the data API", which was both vague and verbose. To keep things as specific as possible, I started to use the codename `DEX` for the Habitat data layer. It has two possible meanings:
+* The data inDEX: `DEX` is like an index or directory for all the data available on the Habitat network. It provides an API that lets you find exactly what data is available, and how to get it.
+* Data EXchange: `DEX` supports interoperable data exchange across communities, applications, and users.
+
+This name is subject to change in the future. But for now, we need to call it something.
+
+## What is DEX?
+DEX is an entrypoint for programs and humans to access all the data on Habitat. 
+
+DEX is necessary to resolve the conflict between two of Habitat's conflicting objectives:
+* Maximum interoperability between applications
+* Support for 3rd party components integrated into all parts of Habitat (specifically datastores in this case)
+
+The wide range of (future) available datastores makes interoperability hard. DEX provides a generic method for accessing data spread across different datastores and communities through a single API. In addition, through `DEX Data Sources`, 
+
+## Universal Data Address Space
+In DEX, data is organized in a universal address space that accounts for different communities, datastores, and namespaces. Data addresses have the following format:
+
+`/dex/<community_id>/<datastore_id>/<namespace>/<content_identifier>`
+
+### Communities
+The highest division of data in Habitat is by community ID. TODO link to the community doc when it exists
+
+### Datastores
+Habitat supports many different datastores. Each type of datastore has a unique identifier, which is usually the conventional name of the datastore plus a version specifier. For example, Postgres 13 might have a datastore identifier of `postgres13`.
+
+### Namespaces
+Namespaces allows for data to be grouped across datastores. This is useful for permissions and authorization, as well as simply giving a high level grouping for data that can be utilized by applications. The namespace has analogues in the different types of datastores. In a RDBMS, a namespace corresponds roughly to a database, while in a filesystem a namespace might correspond to a disk partition or drive. The most common use of namespaces is to group data written by the same application to different datastrores. Every datasource must partition data by namespace.
+
+Different applications can write to the same namespace, but they must agree on what the namespace is for `TODO figure out how to prevent conflicts`. 
```

Comment:
I wonder if instead of applications writing to the same namespace, there's like a different conception of who writes. so maybe there's some small "modules" and all they do is receive a request for writing data & write to it - so for e.g. there's like a location module and if applications want to write to location data it goes through that. then the conflicts are resolved there. let's talk more in meeting